File: python/infinicore/ops/symm.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def symm(
    a: Tensor,
    b: Tensor,
    alpha: Tensor,
    beta: Tensor,
    out: Tensor,
    *,
    side: int = 0,
    uplo: int = 0,
):
    _infinicore.symm_(
        a._underlying,
        b._underlying,
        alpha._underlying,
        beta._underlying,
        out._underlying,
        side,
        uplo,
    )

    return out

File: src/infinicore/ops/symm/symm.cc
#include "infinicore/ops/symm.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(Symm);

Symm::Symm(const Tensor &a, const Tensor &b, const Tensor &alpha, const Tensor &beta, Tensor c, int side, int uplo) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(a, b, alpha, beta, c);
    INFINICORE_GRAPH_OP_DISPATCH(c->device().getType(), a, b, alpha, beta, c, side, uplo);
}

void Symm::execute(const Tensor &a, const Tensor &b, const Tensor &alpha, const Tensor &beta, Tensor c, int side, int uplo) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(Symm, a, b, alpha, beta, c, side, uplo);
}

void symm_(const Tensor &a, const Tensor &b, const Tensor &alpha, const Tensor &beta, Tensor c, int side, int uplo) {
    Symm::execute(a, b, alpha, beta, c, side, uplo);
}

} // namespace infinicore::op
